packages/eslint-plugin/docs/rules/typedef.md

```diff
@@ -0,0 +1,364 @@
+> Instead of enabling `typedef`, it is generally recommended to use the `--noImplicitAny` and/or `--strictPropertyInitialization` compiler options to enforce type annotations only when useful.
+
+## Rule Details
+
+This rule can enforce type annotations in locations regardless of whether they're required.
+This is typically used to maintain consistency for element types that sometimes require them.
+
+## Options
+
+This rule has an object option that may receive any of the following as booleans:
+
+- `"arrayDestructuring"`
+- `"arrowCallSignature"`: `true` by default
+- `"arrowParameter"`: `true` by default
+- `"callSignature"`: `true` by default
+- `"memberVariableDeclaration"`: `true` by default
+- `"objectDestructuring"`
+- `"parameter"`: `true` by default
+- `"propertyDeclaration"`: `true` by default
+- `"variableDeclaration"`
+
+For example, with the following configuration:
+
+```json
+{
+  "rules": {
+    "typedef": [
+      "error",
+      {
+        "arrowParameter": false,
+        "variableDeclaration": true
+      }
+    ]
+  }
+}
+```
+
+- Type annotations on arrow function parameters are not required
+- Type annotations on variables are required
+- Options otherwise adhere to the defaults
+
+### arrayDestructuring
+
+Whether to enforce type annotations on variables declared using array destructuring.
+
+Examples of **incorrect** code with `{ "arrayDestructuring": true }`:
+
+```ts
+const [a] = [1];
+const [b, c] = [1, 2];
+```
+
+Examples of **correct** code with `{ "arrayDestructuring": true }`:
+
+```ts
+const [a]: number[] = [1];
+const [b]: [number] = [2];
+const [c, d]: [boolean, string] = [true, 'text'];
+```
+
+### arrowCallSignature
+
+Whether to enforce type annotations for return types of arrow functions.
+
+Examples of **incorrect** code with `{ "arrowCallSignature": true }`:
+
+```ts
+const returnsVoid = () => {};
+const returnsString = () => 'text';
+
+['hello', 'world'].map(text => text.length);
+
+const mapper = {
+  map: text => text + '...',
+};
+```
+
+Examples of **correct** code with `{ "arrowCallSignature": true }`:
+
+```ts
+const returnsVoid = (): void => {};
+const returnsString = (): string => 'text';
+
+['hello', 'world'].map((text): number => text.length);
+
+const mapper = {
+  map: (text): string => text + '...',
+};
+```
+
+### arrowParameter
+
+Whether to enforce type annotations for parameters of arrow functions.
+
+Examples of **incorrect** code with `{ "arrowParameter": true }`:
+
+```ts
+const logsSize = size => console.log(size);
+
+['hello', 'world'].map(text => text.length);
+
+const mapper = {
+  map: text => text + '...',
+};
+```
+
+Examples of **correct** code with `{ "arrowParameter": true }`:
+
+```ts
+const logsSize = (size: number) => console.log(text);
+
+['hello', 'world'].map((text: string) => text.length);
+
+const mapper = {
+  map: (text: string) => text + '...',
+};
+```
+
+### callSignature
+
+Whether to enforce type annotations for return types of functions and methods.
```

Comment:
as per `arrowCallSignature` option

Reply:
Err, sorry, I don't follow - is that suggested text to add here?

Reply:
sorry - I meant refer to the discussion on `arrowCallSignature`.